File: rate_limit.go
package main

// Should rate limit
// 4 cases
//   1: key not in cache (create new entry, count = 1, return false)
//   2: key in cache, not expired, count < max (increment count, return false)
//   3. key in cache, not expired, count >= max (increment count, return true)
//   4. key in cache, expired (create new entry, count = 1, return false)

import (
	"fmt"
	"time"
)

type entry struct {
	endTime int // Window end time
	counter int // Counter
}

var cache map[string]entry

func main() {
	cache = make(map[string]entry)

	exerciseRateLimiter(5, 2, 120)

	fmt.Println("")
}

func exerciseRateLimiter(count int, interval int, maxCount int) {
	for i := 0; i < count; i++ {
		for {
			shouldLimit := rateLimit("device_info", interval, maxCount)
			if shouldLimit {
				// Rate limited
				fmt.Println("")
				if i < count-1 {
					for {
						time.Sleep(5 * time.Millisecond)
						shouldLimit = rateLimit("device_info", interval, maxCount)
						if !shouldLimit {
							fmt.Print(".")
							break
						}
					}
				}
				break
			} else {
				// Not rate limited.
				fmt.Print(".")
			}
		}
	}
}

func rateLimit(key string, interval int, maxCount int) bool {
	t := int(time.Now().Unix())
	e, ok := cache[key]
	if ok {
		// Key in cache
		if t >= e.endTime {
			// Case 4
			cache[key] = entry{endTime: t + interval, counter: 1}
			return false
		} else {
			if e.counter >= maxCount {
				// Case 3
				return true
			} else {
				// Case 2
				e.counter++
				cache[key] = e
				return false
			}
		}
	} else {
		// Case 1
		cache[key] = entry{endTime: t + interval, counter: 1}
		return false
	}
}

File: rate_limit.lua
--[[
Should rate limit
4 cases
  1: key not in cache (create new entry, count = 1, return false)
  2: key in cache, not expired, count < max (increment count, return false)
  3. key in cache, not expired, count >= max (return true)
  4. key in cache, expired (create new entry, count = 1, return false)
--]]


cache = {}


function unixTime()
    return os.time(os.date("!*t"))
end


function rateLimit(key, interval, maxCount)
    local t = unixTime()
    local e = cache[key]
    if e == nil then
        -- Not in cache
        -- Case 1
        cache[key] = {endTime = t + interval, count = 1}
        return false
    end

    -- In cache
    if t < e.endTime then
        -- Not expired
        if e.count < maxCount then
            -- In cache, not expired, count < max
            -- Case 2
            e.count = e.count + 1
            cache[key] = e
            return false
        end

        -- In cache, not expired, count >= max
        -- Case 3
        return true
    end

    -- Expired
    -- Case 4
    cache[key] = {endTime = t + interval, count = 1}
    return false
end


function test(key, interval, maxCount)
    -- false
    print(rateLimit(key, interval, maxCount))
    print(rateLimit(key, interval, maxCount))
    print(rateLimit(key, interval, maxCount))

    -- true
    print(rateLimit(key, interval, maxCount))
end


test("key1", 5, 3)